### PR TITLE
[FEATURE] Added support for Swift 4's Codable

### PR DESCRIPTION
**Pull request by samdods**
*NOTE: I'm interested to hear other people's thoughts on how I've tackled the problem of nesting within the response. Open to other suggestions!*

This change adds the following protocols:

```
JSONDecodableResourceType
PropertyListDecodableResourceType
NetworkJSONDecodableResourceType
NetworkPropertyListDecodableResourceType
```

Note that we don't have to provide the equivalent "array" resource types, like we do for typical JSON resources. This is because if `TypeX` is `Decodable` then `[TypeX]` is also `Decodable`. We would just use the `JSONDecodableResourceType` and specify the typealias `Model = [TypeX]`.

### CODE REVIEW

File: Sources/TABResourceLoader/Protocols/Resource types/NetworkJSONDecodableResourceType.swift [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+//  THE SOFTWARE.
+
+import Foundation
+
+/// Defines a resource that can be fetched from a network where the root type is Decodable JSON
+public protocol NetworkJSONDecodableResourceType: NetworkResourceType, JSONDecodableResourceType {}
+
+public extension NetworkJSONDecodableResourceType {
+  var httpHeaderFields: [String: String]? {
+    return ["Content-Type": "application/json"]
```
Comment (lucianomarisi) [inline Sources/TABResourceLoader/Protocols/Resource types/NetworkJSONDecodableResourceType.swift:36]: Probably a good idea to add tests to assert these headers similar to `NetworkJSONArrayResourceTypeTests`

**Review by lucianomarisi — COMMENTED** [review]
Looks great!
- It's worth updating the README to document that `Decodable` types are now supported
- Not convinced having a property list resource is so useful, aside from loading from disk, I'm not sure there is a common usecase to send property lists over the network. Adding this means that there is more code to maintain in the future without any real benefits.

File: Sources/TABResourceLoader/Protocols/Resource types/NetworkPropertyListDecodableResourceType.swift [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+//  THE SOFTWARE.
+
+import Foundation
+
+/// Defines a resource that can be fetched from a network where the root type is a Decodable Property List
+public protocol NetworkPropertyListDecodableResourceType: NetworkResourceType, PropertyListDecodableResourceType {}
+
+public extension NetworkPropertyListDecodableResourceType {
+  var httpHeaderFields: [String: String]? {
+    return ["Content-Type": "application/json"]
```
Comment (samdods) [inline Sources/TABResourceLoader/Protocols/Resource types/NetworkPropertyListDecodableResourceType.swift]: just realised this isn't correct.

**Comment by samdods** [comment]
good point on the README.

not sure about the PropertyList. i kinda think it's worth keeping it... but TABResourceLoader doesn't support loading from disk does it. so the resource type is a bit misplaced.

in fact, that's just made me think, all the resource types are a bit useless without the corresponding NetworkResourceType conformance. because there is no service that will load from anything other than the network.

File: Example/Models/Codable examples/ProductNestedInResponseResource.swift [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+//  Copyright © 2017 Luciano Marisi. All rights reserved.
+//
+
+import Foundation
+import TABResourceLoader
+
+struct ProductNestedInResponseResource: NetworkJSONDecodableResourceType {
+  typealias Model = Product
+
+  // swiftlint:disable nesting
```
Comment (jsanderson44) [inline Example/Models/Codable examples/ProductNestedInResponseResource.swift]: Is it worth just disabling this globally in `swiftlint.yml`?

File: Example/ViewController.swift [hunk 4]
```diff
@@ -62,6 +62,21 @@ class ViewController: UIViewController {
     }
   }
 
+  func fetchDecodable() {
```
Comment (jsanderson44) [inline Example/ViewController.swift]: If we're being consistent should this technically be `fetchDecodableExample()`? (based on other functions in this class)

File: Tests/TABResourceLoaderTests/Protocols/JSONDecodableResourceTypeTests.swift [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+import XCTest
+@testable import TABResourceLoader
+
+class JSONDecodableResourceTypeTests: XCTestCase {
+
+  // MARK: - Valid objects
+
+  func test_validDecodableObject() throws {
+    let mockResource = MockJSONDecodableResource()
+    let data = try """
```
Comment (jsanderson44) [inline Tests/TABResourceLoaderTests/Protocols/JSONDecodableResourceTypeTests.swift:18]: Nice use of `"""` 👍

**Comment by KaneCheshire** [comment]
[Add support for Swift 4's Codable to TABResourceLoader](https://trello.com/c/hvddZQqB/2-add-support-for-swift-4s-codable-to-tabresourceloader)

**Review by jsanderson44 — APPROVED** [review]

**Review by KaneCheshire — APPROVED** [review]

**Pull request merged**